stdlib/public/core/Filter.swift

```diff
@@ -222,9 +230,13 @@ public struct LazyFilterCollection<
   // TODO: swift-3-indexing-model - add docs
   @warn_unused_result
   public func advance(i: Index, by n: IndexDistance, limit: Index) -> Index {
+    _precondition(n >= 0,
+      "Only BidirectionalCollections can be advanced by a negative amount")
+    // TODO: swift-3-indexing-model: _failEarlyRangeCheck i?
+
     var index = i.base
-    for _ in 0..<n where index != _base.endIndex && index != limit.base {
-      index = _nextFiltered(index)
+    for _ in 0..<n {
+      _nextFilteredInPlace(&index, limit: limit.base)
```

Comment:
Once we hit the limit, we should stop looping.  `_nextFilteredInPlace` has that information, it should probably return it.